DaaS/V2/Diagnostics/PredefinedValidators.cs

```diff
@@ -0,0 +1,268 @@
+        }
+
+        public bool JavaFlightRecorderValidator(out string AdditionalError)
+        {
+            return CheckJavaProcessAndTools("jcmd.exe", out AdditionalError);
+        }
+
+        private string GetJavaFolderPathFromConfig(out bool pathInConfigNotJavaExe)
+        {
+            pathInConfigNotJavaExe = false;
+            string javaExePath = string.Empty;
+            string webConfigPath = EnvironmentVariables.WebConfigFilePath;
+            if (File.Exists(webConfigPath))
+            {
+                var xdocument = XDocument.Load(webConfigPath);
+                var httpPlatform = xdocument.Descendants().Where(p => p.Name.LocalName == "httpPlatform").FirstOrDefault();
+                if (httpPlatform != null && httpPlatform.Attribute("processPath") != null)
+                {
+                    var processPath = (string)httpPlatform.Attribute("processPath");
+                    processPath = processPath.Contains("%") ? Environment.ExpandEnvironmentVariables(processPath) : processPath;
+                    if (processPath.EndsWith("java.exe", StringComparison.OrdinalIgnoreCase))
+                    {
+                        javaExePath = Path.GetDirectoryName(processPath);
+                    }
+                    else
+                    {
+                        pathInConfigNotJavaExe = true;
+                    }
+                }
+            }
+            return javaExePath;
+        }
+
+        private bool CheckJavaProcessAndTools(string toolName, out string AdditionalError)
+        {
+            AdditionalError = string.Empty;
+            var javaProcess = Process.GetProcessesByName("java").FirstOrDefault();
+            if (javaProcess == null)
+            {
+                return false;
+            }
+
+            var javaFolderPath = GetJavaFolderPathFromConfig(out bool pathInConfigNotJavaExe);
+            if (!string.IsNullOrWhiteSpace(javaFolderPath))
+            {
+                var javaToolPath = Path.Combine(javaFolderPath, toolName);
+                if (File.Exists(javaToolPath))
+                {
+                    return true;
+                }
+                else
+                {
+                    var jcmdPath = Path.Combine(javaFolderPath, "jcmd.exe");
+                    AdditionalError = GetAdditionalError(toolName, jcmdPath);
+                    return false;
+                }
+            }
+            else
+            {
+                //
+                // Try to get the path of rt.jar by running KuduHandles on java.exe
+                //
+
+                var rtJarHandle = GetJarFileHandle(javaProcess.Id);
+                if (!string.IsNullOrWhiteSpace(rtJarHandle) && rtJarHandle.Length > 0)
+                {
+                    var parentPath = rtJarHandle.Replace(@"\jre\lib\rt.jar", "").Replace("c:", "d:");
+                    var javaToolPath = Path.Combine(parentPath, "bin", toolName);
```

Comment:
I see a pattern here in the code where you are checking that if a file exists (for the tool), you return true else you have an error message and return false. Do you feel repetitive code can be reduced by having a method something like :

`bool CheckToolExists(string path, out errorMsg);`

You can then simply use it like this:

`return CheckToolExists(javaToolPath, out AdditionalError);`

Reply:
Nice suggestion. Implemented!